Add logo block

aristath opened the pull request:
Trac ticket: https://core.trac.wordpress.org/ticket/53247

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

youknowriad commented:
Can we enable the site-logo block in the same patch and add the required code? (I think to do so, there's a couple PHP files/webpack config to change and run `npm run build:dev`)

youknowriad commented:
Take the PostAuthor block as an example:

see this line https://github.com/WordPress/wordpress-develop/pull/1273/files#diff-c62beff71446448ff952e38805c8ae7652f4e2717d3ca41b1d9cd9963bd82464L17

and 

 https://github.com/WordPress/wordpress-develop/pull/1273/files#diff-47f938308062dccd6b66f867fd9030b08c2f582b8e2e28351b999d84360d871dL116

aristath commented:
Thank you @youknowriad :+1: 
Added the block and fixed a related phpunit test.

youknowriad commented:
@aristath looks like there are some failing phpunit tests here, any idea?

aristath commented:
Fixed the failing tests

ocean90 commented on src/wp-includes/blocks/site-logo.php:
```diff
@@ -0,0 +1,75 @@
+
+	$wrapper_attributes = get_block_wrapper_attributes( array( 'class' => implode( ' ', $classnames ) ) );
+	$html               = sprintf( '<div %s>%s</div>', $wrapper_attributes, $custom_logo );
+	remove_filter( 'wp_get_attachment_image_src', $adjust_width_height_filter );
```
Shouldn't this be right below the `get_custom_logo();` call so it also gets removed if `$custom_logo` is empty?

aristath replied:
Good point. Pushed here, I'll backport the tweak to Gutenberg too :+1:

TimothyBJacobs commented:
At a minimum the name for the theme mods option in the REST API should just be `theme_mods`. It also should be using `set_theme_mod` and probably `get_theme_mod`. But again, putting that in this endpoint is entirely incorrect. The capability is incorrect, it is using `manage_options` instead of `edit_theme_options` and conceptually is a mismatch as the theme mods API is an abstraction over the options API.

Theme mods should either be added to the themes endpoint, or more ideally have an endpoint to interact with registered settings in the customizer.

This is not suitable to be shipped in Core. If you aren't interested in making those endpoints, then I'd recommend looking at calling the customizer admin-ajax API. Ex: https://github.com/WordPress/gutenberg/blob/854a31f3edee98e8b16ef5f2d32abcdaa8b395b0/packages/edit-navigation/src/store/actions.js#L195

This also seems to have a different change of adding the `stylesheet` option to the settings endpoint. I'm not sure why this is here, but it can't ship. If you need the active theme get it from the themes endpoint. If you need to change themes, you can't just do that by updating the option. Theme switching needs to be added to the themes endpoint.

aristath commented:
The logo block was updated in https://github.com/WordPress/gutenberg/pull/32229. Once a new minor Gutenberg release goes out I will update this PR with the new implementation (`npm run build:dev` doesn't pull from Gutenberg's trunk)

aristath commented:
For the sake of this PR I updated `@wordpress/block-library` to v3.2.2, cleaned-up previous files and updated the ones relevant to the site-logo block.
Tests are currently failing, but it _should_ be OK once the `block-library` package gets updated in the `master` branch and this gets rebased.

youknowriad commented:
Packages are now up to date on trunk, would you mind refreshing this PR?

aristath commented:
looking into the failing tests now :+1:

aristath commented:
All tests now pass. I needed to tweak the sync hook a bit, and a PR backporting these changes to Gutenberg was submitted in  https://github.com/WordPress/gutenberg/pull/32370

youknowriad commented:
Ok makes sense. I guess this means we should wait for the next package update to land this PR to avoid having `npm run build` override your changes.

aristath commented on src/wp-includes/blocks/site-logo.php:
```diff
@@ -0,0 +1,156 @@
+	}
+}
+
+$theme = get_option( 'stylesheet' );
+add_action( "update_option_theme_mods_$theme", '_sync_custom_logo_to_site_logo', 10, 2 );
```
backporting from https://github.com/WordPress/gutenberg/pull/32370#pullrequestreview-673677439

```suggestion
add_action( 'update_option_theme_mods_' . get_option( 'stylesheet' ), '_sync_custom_logo_to_site_logo', 10, 2 );
```

aristath commented:
Updated, should be OK to merge now :+1:

youknowriad commented:
committed in https://core.trac.wordpress.org/changeset/51091
